consensus/src/lib.rs

```diff
@@ -1,27 +1,1109 @@
-// use blockchain_storage::BlockchainStorage;
-// use common::chain::config::ChainConfig;
-// use common::chain::block::Block;
+// Copyright (c) 2021 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://spdx.org/licenses/MIT
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author(s): S. Afach, A. Sinitsyn
 
+use crate::orphan_blocks::{OrphanAddError, OrphanBlocksPool};
+use blockchain_storage::BlockchainStorageRead;
+use blockchain_storage::BlockchainStorageWrite;
+use blockchain_storage::TransactionRw;
+use blockchain_storage::Transactional;
+use common::chain::block::block_index::BlockIndex;
+use common::chain::block::{calculate_tx_merkle_root, calculate_witness_merkle_root, Block};
+use common::chain::config::ChainConfig;
+use common::chain::{
+    OutPoint, OutPointSourceId, OutputSpentState, SpendablePosition, Spender, Transaction,
+    TxMainChainIndex, TxMainChainPosition,
+};
+use common::chain::{SpendError, TxMainChainIndexError};
+use common::primitives::{time, Amount, BlockHeight, Id, Idable};
+use thiserror::Error;
 mod orphan_blocks;
-// use orphan_blocks::OrphanBlocks;
+use parity_scale_codec::Encode;
+use std::collections::BTreeMap;
 
-// struct Consensus<S: BlockchainStorage> {
-//     chain_config: ChainConfig,
-//     blockchain_storage: S,
-//     orphan_blocks: OrphanBlocks,
-// }
+#[allow(dead_code)]
+#[derive(Error, Debug, PartialEq, Eq)]
+pub enum BlockError {
+    #[error("Unknown error")]
+    Unknown,
+    // Orphan block
+    #[error("Orphan")]
+    Orphan,
+    #[error("Invalid block height `{0}`")]
+    InvalidBlockHeight(BlockHeight),
+    #[error("The previous block invalid")]
+    PrevBlockInvalid,
+    #[error("The storage cause failure `{0}`")]
+    StorageFailure(blockchain_storage::Error),
+    #[error("The block not found")]
+    NotFound,
+    #[error("Invalid block source")]
+    InvalidBlockSource,
+    // To be expanded
+}
+
+impl From<blockchain_storage::Error> for BlockError {
+    fn from(_err: blockchain_storage::Error) -> Self {
+        // On storage level called err.recoverable(), if an error is unrecoverable then it calls panic!
+        // We don't need to cause panic here
+        BlockError::Unknown
+    }
```

Comment:
Lots of errors have to be added and classified as recoverable or irrecoverable.

Reply:
@iljakuklic @TheQuantumPhysicist Let's make an issue and resolve it in a next PR because some people blocked until this review will finish. What do you think, guys?

Reply:
I am fine with that

Reply:
Done https://github.com/mintlayer/mintlayer-core/issues/126#issue-1187614914